File: src/test/java/com/automation/um/pages/UsersPage.java
package com.automation.um.pages;

import com.automation.um.utils.WaitUtils;
import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public class UsersPage extends BasePage {
  public UsersPage(WebDriver driver) {
    super(driver);
  }

  public void openUsers() {
    open("/users");
    waitForTableReady();
  }

  public void waitForTableReady() {
    WaitUtils.invisible(driver, testId("users-row-skeleton"));
    if (driver.findElements(testId("users-table")).isEmpty()) {
      WaitUtils.visible(driver, testId("users-empty-add"));
    } else {
      el("users-table");
    }
  }

  public void search(String query) {
    fillInput("users-search", query);
    WaitUtils.debounce();
    waitForTableReady();
  }

  public void filterByRole(String role) {
    selectOptionByValue("users-filter-role", role);
    waitForTableReady();
  }

  public void filterByStatus(String status) {
    selectOptionByValue("users-filter-status", status);
    waitForTableReady();
  }

  public void setPageSize(int size) {
    selectOptionByValue("users-page-size", String.valueOf(size));
    waitForTableReady();
  }

  public void sortBy(String column) {
    el("sort-" + column).click();
    waitForTableReady();
  }

  public void refresh() {
    el("users-refresh").click();
    waitForTableReady();
  }

  public void openCreateModal() {
    el("users-add").click();
    el("user-form-submit");
  }

  public void openEditForUser(String userId) {
    el("user-edit-" + userId).click();
    el("user-form-submit");
  }

  public void openDetailsForUser(String userId) {
    el("user-view-" + userId).click();
    el("details-close");
  }

  public void deleteUser(String userId) {
    el("user-delete-" + userId).click();
  }

  public void selectUser(String userId) {
    WebElement checkbox = el("user-select-" + userId);
    if (!checkbox.isSelected()) {
      checkbox.click();
    }
  }

  public void selectAllVisible() {
    WebElement checkbox = el("users-select-all");
    if (!checkbox.isSelected()) {
      checkbox.click();
    }
  }

  public void bulkDelete() {
    el("users-bulk-delete").click();
  }

  public void nextPage() {
    el("users-next").click();
    waitForTableReady();
  }

  public void previousPage() {
    el("users-prev").click();
    waitForTableReady();
  }

  public int visibleRowCount() {
    return driver.findElements(By.cssSelector("[data-testid^='user-row-']")).size();
  }

  public String firstUserId() {
    List<WebElement> rows = driver.findElements(By.cssSelector("[data-testid^='user-row-']"));
    if (rows.isEmpty()) return null;
    String testId = rows.get(0).getAttribute("data-testid");
    return testId == null ? null : testId.replace("user-row-", "");
  }

  public boolean rowExistsForEmail(String email) {
    return driver.getPageSource().contains(email);
  }

  public boolean rowExists(String userId) {
    return isVisible("user-row-" + userId);
  }

  public String selectedCountText() {
    return el("users-selected-count").getText();
  }

  public String pageText() {
    return el("users-page").getText();
  }

  public boolean isNextEnabled() {
    return el("users-next").isEnabled();
  }
}

File: playwright/pages/UsersPage.js
import { BasePage } from "./BasePage.js";
import * as WaitUtils from "../utils/waitUtils.js";

export class UsersPage extends BasePage {
  async openUsers() {
    await this.open("/users");
    await this.waitForTableReady();
  }

  async waitForTableReady() {
    await WaitUtils.invisible(this.page, "users-row-skeleton").catch(() => {});
    const hasTable = await this.testId("users-table").count();
    if (hasTable === 0) {
      await this.el("users-empty-add");
    } else {
      await this.el("users-table");
    }
  }

  async search(query) {
    await this.fillInput("users-search", query);
    await WaitUtils.debounce();
    await this.waitForTableReady();
  }

  async filterByRole(role) {
    await this.selectOptionByValue("users-filter-role", role);
    await this.waitForTableReady();
  }

  async filterByStatus(status) {
    await this.selectOptionByValue("users-filter-status", status);
    await this.waitForTableReady();
  }

  async setPageSize(size) {
    await this.selectOptionByValue("users-page-size", String(size));
    await this.waitForTableReady();
  }

  async sortBy(column) {
    await this.testId(`sort-${column}`).click();
    await this.waitForTableReady();
  }

  async refresh() {
    await this.testId("users-refresh").click();
    await this.waitForTableReady();
  }

  async openCreateModal() {
    await this.testId("users-add").click();
    await this.el("user-form-submit");
  }

  async openEditForUser(userId) {
    await this.testId(`user-edit-${userId}`).click();
    await this.el("user-form-submit");
  }

  async openDetailsForUser(userId) {
    await this.testId(`user-view-${userId}`).click();
    await this.el("details-close");
  }

  async deleteUser(userId) {
    await this.testId(`user-delete-${userId}`).click();
  }

  async selectUser(userId) {
    const checkbox = this.testId(`user-select-${userId}`);
    if (!(await checkbox.isChecked())) {
      await checkbox.check();
    }
  }

  async selectAllVisible() {
    const checkbox = this.testId("users-select-all");
    if (!(await checkbox.isChecked())) {
      await checkbox.check();
    }
  }

  async bulkDelete() {
    await this.testId("users-bulk-delete").click();
  }

  async nextPage() {
    await this.testId("users-next").click();
    await this.waitForTableReady();
  }

  async previousPage() {
    await this.testId("users-prev").click();
    await this.waitForTableReady();
  }

  async visibleRowCount() {
    return this.page.locator("[data-testid^='user-row-']").count();
  }

  async firstUserId() {
    const testId = await this.page.locator("[data-testid^='user-row-']").first().getAttribute("data-testid");
    return testId?.replace("user-row-", "") ?? null;
  }

  async rowExistsForEmail(email) {
    const content = await this.page.content();
    return content.includes(email);
  }

  async rowExists(userId) {
    return this.isVisible(`user-row-${userId}`);
  }

  async selectedCountText() {
    return (await this.testId("users-selected-count").textContent())?.trim() ?? "";
  }

  async pageText() {
    return (await this.testId("users-page").textContent())?.trim() ?? "";
  }

  async isNextEnabled() {
    return this.testId("users-next").isEnabled();
  }
}
